bitcoin/zcashd/zcashprices.go

```diff
@@ -0,0 +1,283 @@
+	z.fetchCurrentRates()
+	z.Lock()
+	defer z.Unlock()
+	price, ok := z.cache[currencyCode]
+	if !ok {
+		return 0, errors.New("Currency not tracked")
+	}
+	return price, nil
+}
+
+func (z *ZcashPriceFetcher) GetAllRates() (map[string]float64, error) {
+	z.Lock()
+	defer z.Unlock()
+	return z.cache, nil
+}
+
+func (z *ZcashPriceFetcher) UnitsPerCoin() int {
+	return exchange.SatoshiPerBTC
+}
+
+func (z *ZcashPriceFetcher) fetchCurrentRates() error {
+	z.Lock()
+	defer z.Unlock()
+	for _, provider := range z.providers {
+		err := provider.fetch()
+		if err == nil {
+			return nil
+		}
+	}
+	log.Error("Failed to fetch zcash exchange rates")
+	return errors.New("All exchange rate API queries failed")
+}
+
+func (z *ZcashPriceFetcher) run() {
+	z.fetchCurrentRates()
+	ticker := time.NewTicker(time.Minute * 15)
+	for range ticker.C {
+		z.fetchCurrentRates()
+	}
+}
+
+func (provider *ExchangeRateProvider) fetch() (err error) {
+	if len(provider.fetchUrl) == 0 {
+		err = errors.New("Provider has no fetchUrl")
+		return err
+	}
+	resp, err := provider.client.Get(provider.fetchUrl)
+	if err != nil {
+		log.Error("Failed to fetch from "+provider.fetchUrl, err)
+		return err
+	}
+	decoder := json.NewDecoder(resp.Body)
+	var dataMap interface{}
+	err = decoder.Decode(&dataMap)
+	if err != nil {
+		log.Error("Failed to decode JSON from "+provider.fetchUrl, err)
+		return err
+	}
+	return provider.decoder.decode(dataMap, provider.cache, provider.bitcoinProvider)
+}
+
+func (b KrakenDecoder) decode(dat interface{}, cache map[string]float64, bp *exchange.BitcoinPriceFetcher) (err error) {
+	rates, err := bp.GetAllRates()
+	if err != nil {
+		return err
+	}
+	obj, ok := dat.(map[string]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	result, ok := obj["result"]
+	if !ok {
+		return errors.New("KrakenDecoder: field `result` not found")
+	}
+	resultMap, ok := result.(map[string]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	pair, ok := resultMap["BCHXBT"]
+	if !ok {
+		return errors.New("KrakenDecoder: field `BCHXBT` not found")
+	}
+	pairMap, ok := pair.(map[string]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	c, ok := pairMap["c"]
+	if !ok {
+		return errors.New("KrakenDecoder: field `c` not found")
+	}
+	cList, ok := c.([]interface{})
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	rateStr, ok := cList[0].(string)
+	if !ok {
+		return errors.New("KrackenDecoder type assertion failure")
+	}
+	price, err := strconv.ParseFloat(rateStr, 64)
+	if err != nil {
+		return err
+	}
+	rate := price
+
+	if rate == 0 {
+		return errors.New("Bitcoin-BitcoinCash price data not available")
+	}
+	for k, v := range rates {
+		cache[k] = v * rate
+	}
+	return nil
+}
+
+func (b BitfinexDecoder) decode(dat interface{}, cache map[string]float64, bp *exchange.BitcoinPriceFetcher) (err error) {
```

Comment:
Again, I think we should try using a struct here over the assertions.

Reply:
What's the benefit to doing it that way. I just converted one and it ended up being about the same number of lines of code.

Reply:
It won't ever panic.

Reply:
well this shouldn't panic either because each type assertion is checked.